File: csrc/block.cpp
#include "block.h"
#include <stdexcept>

Block::Block(int block_id_,
             const std::string& compare_mode_,
             int external_blocks_num_)
    : block_id(block_id_),
      ref_count(0),
      hash(-1),
      hit_count(0),
      prefix_position(0),
      prefix_depth(0),
      external_blocks_num(external_blocks_num_),
      compare_mode(compare_mode_)
{
    if (compare_mode != "asc" && compare_mode != "desc" && compare_mode != "local_first") {
        throw std::invalid_argument("compare_mode must be 'asc', 'desc' or 'local_first'");
    }
}

void Block::update(int _hash, const std::vector<int>& _token_ids) {
    hash = _hash;
    token_ids = _token_ids;
}

void Block::reset() {
    ref_count = 1;
    hash = -1;
    token_ids.clear();
}

int Block::dist() const {
    return prefix_depth - prefix_position;
}

bool Block::operator<(const Block& other) const {
    if (hash == -1 && other.hash != -1)
        return true;
    if (hash != -1 && other.hash == -1)
        return false;
    if (hit_count != other.hit_count)
        return hit_count < other.hit_count;
    if (dist() != other.dist())
        return dist() < other.dist();

    if (compare_mode == "asc")
        return block_id < other.block_id;

    if (compare_mode == "desc")
        return block_id > other.block_id;

    if (compare_mode == "local_first") {
        if (block_id < external_blocks_num && other.block_id < external_blocks_num)
            return block_id < other.block_id;
        if (block_id >= external_blocks_num && other.block_id < external_blocks_num)
            return true;
        if (block_id < external_blocks_num && other.block_id >= external_blocks_num)
            return false;
        if (block_id >= external_blocks_num && other.block_id >= external_blocks_num)
            return block_id < other.block_id;
    }
    return false;
}

File: swiftcache/engine/block_manager/block.py

class Block:

    def __init__(self, block_id,compare_mode="asc",external_blocks_num = 0):
        self.block_id = block_id
        self.ref_count = 0
        self.hash = -1
        self.token_ids = []
        self.hit_count = 0
        self.prefix_position = 0
        self.prefix_depth = 0
        # self.block_hash
        self.external_blocks_num = external_blocks_num
        assert compare_mode in ['asc','desc','local_first']
        self.compare_mode = compare_mode
    def update(self, _hash: int, token_ids: list[int]):
        self.hash = _hash
        self.token_ids = token_ids

    def reset(self):
        self.ref_count = 1
        self.hash = -1
        self.token_ids = []
    
    @property
    def dist(self):
        return self.prefix_depth - self.prefix_position

    def __lt__(self, other: "Block") -> bool:
        # 规则1：hit_count 小的优先级高
        if self.hash == -1 and other.hash != -1:
            return True
        if self.hash != -1 and other.hash == -1:
            return False
        if self.hit_count != other.hit_count:
            return self.hit_count < other.hit_count
        # if self.prefix_position != other.prefix_position:
        #     return self.prefix_position > other.prefix_position
        # if self.prefix_depth != other.prefix_depth:
        #     return self.prefix_depth < other.prefix_depth
        if self.dist != other.dist:
            return self.dist < other.dist

        if self.compare_mode == 'asc':
            return self.block_id < other.block_id
        if self.compare_mode == 'desc':
            return self.block_id > other.block_id
        if self.compare_mode == 'local_first':
            if self.block_id < self.external_blocks_num and other.block_id < self.external_blocks_num :
                return self.block_id < other.block_id
            if self.block_id >= self.external_blocks_num and other.block_id < self.external_blocks_num :
                return True
            if self.block_id < self.external_blocks_num and other.block_id >= self.external_blocks_num :
                return False
            if self.block_id >= self.external_blocks_num and other.block_id >= self.external_blocks_num :
                return self.block_id < other.block_id